### PR TITLE
sql/apd: properly rescale values on insert

### LINKED ISSUE #7220: apd: rescale_apd applied even if source and destination have the same scale/precision
### What version of Materialize are you using?

```
v0.8.2-dev (56edf106e)
```

### What was the issue?

``rescale_apd``` appears in the plan even though the source and the destination have the same scale/precision.

Ideally, no operator should be applied in this case as parasitic operators like that tend to prevent various optimizations from kicking in.

At the very least, ```rescale_apd``` should be a no-op of sorts. I tried to follow the call chain but I could not see an early exit if the scale and the precision are identical. 

### Is the issue reproducible? If so, please provide reproduction instructions.

```
create table t1(f1 decimal(5,2));

```
materialize=> EXPLAIN SELECT f1::decimal(5,2) FROM t1;
          Optimized Plan           
-----------------------------------
 %0 =                             +
 | Get materialize.public.t1 (u37)+
 | Map rescale_apd(#0)            +
 | Project (#1)                   +
```

### LINKED ISSUE #7226: apd: Values that have passed through rounding at insert time are not considered equal to identical values
### What version of Materialize are you using?

```
materialized v0.8.2-dev (56edf106e)
```

### What was the issue?

You can define a column as ```DECIMAL|APD(M,N)``` and then insert a value with more than N digits after the decimal sign. This value will be rounded as expected when SELECT-ed but it is not considered equal to an identical value that has not passed through the same rounding process. 

### Is the issue reproducible? If so, please provide reproduction instructions.

```
DROP TABLE IF EXISTS t1 CASCADE;
CREATE TABLE t1 (f1 DECIMAL(5,2));
INSERT INTO t1 VALUES (0.0001);
SELECT f1, f1 = 0 FROM t1;
  f1  | ?column? 
------+----------
 0.00 | f <- so zero is apparently not considered identical to zero
```

Contrast with Postgres:

```
pstoev=# SELECT f1, f1 = 0 FROM t1;
  f1  | ?column? 
------+----------
 0.00 | t
(1 row)
```

It also works as expected with the old decimal type.

### PR DESCRIPTION
**Pull request by sploiselle**
The fixed issue was caused by a failure to rescale APD values on insert (though they did get properly rescaled when read from the tables out to `psql`). The underlying cause of _that_ was that `cast_relation` thought all APD columns were equal irrespective of scale due to its reliance of `ScalarType::eq`, and didn't perform the casts we expected.

Rather than move the pseudo-equal logic from `get_cast` to a helper function, I instead have `cast_relation` call `plan_cast` on every column and throw away the unnecessary casts.

Fixes #7226

Also...
Fixes #7220

### CODE REVIEW

File: src/sql/src/plan/query.rs [hunk 1]
```diff
@@ -689,7 +695,7 @@ fn cast_relation<'a, I>(
     target_types: I,
 ) -> Result<HirRelationExpr, CastRelationError>
 where
-    I: IntoIterator<Item = &'a ScalarType>,
+    I: Iterator<Item = &'a ScalarType>,
```
Comment (benesch) [inline src/sql/src/plan/query.rs:698]: nit, but this looks like it should stay `IntoIterator` and the call to `into_iter()` should go where `target_types` is used: `target_types.into_iter().enumerate()`

File: test/testdrive/apd.td [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+# Note that this test uses an append-only source, so doesn't fully express the
+# semantics of summing APD values because values are non-retractable. # A future
+# iteration of this test will use a more robust source and improve the tests # to
+# more accurately reflect the operation's semantics.
```
Comment (benesch) [inline test/testdrive/apd.td:13]: Minor: funny wrapping

File: src/repr/src/scalar.rs [hunk 3]
```diff
@@ -864,9 +864,13 @@ impl<'a> ScalarType {
...
-// TODO(benesch): the implementations of PartialEq and Hash for ScalarType can
-// be removed when decimal precision is either fixed or removed.
-
+/// NOTE: `ScalarType` equality is complicated by variants that store values.
+/// While we might want to handle stored values in one manner most of the time,
+/// that same approach might be insufficient in some cases.
+///
+/// When assessing unexpected behavior with type planning, it's worthwhile to
+/// see if the default equality is overridden in some other part of the code,
+/// e.g. when deciding how to handle potential casts between types.
```
Comment (benesch) [inline src/repr/src/scalar.rs:873]: This seems good for now but long-term fishy to me. I think we want to eventually change `ScalarType` to have a normal derived `PartialEq` implementation, and then the things that want to be fuzzier about comparisons (e.g. ignoring the APD scale) can do _that_ with non-standard equality methods.

**Review by benesch — APPROVED** [review]

File: src/sql/src/plan/query.rs (hunk 1 above)
Comment (sploiselle) [reply src/sql/src/plan/query.rs:698]: Ah! I had tried a few different permutations of this but not that one. I realize now that the type implements the trait, not that it itself is an iterator.

File: src/repr/src/scalar.rs (hunk 3 above)
Comment (sploiselle) [reply src/repr/src/scalar.rs:873]: Makes sense. Will be easier to shake out once the refactor lands.

**Pull request merged**